### PR TITLE
`Programming exercises`: Add gradle as project type for Java exercises on GitLab/Jenkins setup

### LINKED ISSUE #4841: Add Gradle project type to Java programming exercises on Gitlab/Jenkins Setup
### Is your feature request related to a problem?

Gradle will be added as a new project type for Java programming exercises on the Atlassian Setup within #4781. This could also be interesting for programming exercises for Artemis instances using the GitLab/Jenkins-Setup.

### Describe the solution you'd like

In order to add the option for Gradle projects to the Gitlab/Jenkins-Setup, `JenkinsService.java` has to be extended with the functionality for Gradle builds as well as `JenkinsProgrammingLanguageFeatureService.java` defining the supported language features.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ole-ve**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [X] I implemented the changes with a good performance and prevented too many database calls.
- [X] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite). (changes have no effect)
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
Closes #4841.
In #4781, Gradle has been added as a project type for Java exercises on the Atlassian Setup. Within this PR, the Jenkins build plan creation is extended to also add the option for Gradle on the GitLab/Jenkins-Setup. Including support for
- Static Code Analysis and
- Sequential Test Runs.

### Description
The build plan creation for Java exercises is extended to distinct between Jenkinsfiles for Maven and Gradle projects.

### Steps for Testing
- 1 Instructor
- TS2 (GitLab/Jenkins-Setup)

1. Log in to Artemis
2. Navigate to Course Administration
3. Create Programming Exercises with the following configurations.
- Template Solution Repository Type: Plain or Gradle (does not matter)
- Test Repository Type: Gradle
- SCA: enabled; sequential test runs: disabled OR SCA: disabled, sequential test runs: enabled
4. Confirm that the initial result is created correctly.
5. Push some changes and verify that the results are built correctly.
6. Create a programming exercise with Maven and SCA enabled and confirm that everything works correctly.

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### CODE REVIEW

**Review by dfuchss — APPROVED** [review]
I simply tested whether a gradle project can be created & whether the test results (with static code analysis) are submitted correctly. This works as far as I can see :)

**Review by dfuchss — DISMISSED** [review]

**Review by maximiliansoelch — COMMENTED** [review]
Tested on TS2, creation of the gradle project works as expected. However, I noticed that no build output is available in case of a failed build:
noBuildOutput
 
In contrast, creating a programming exercise with the same settings on TS1 (Atlassian Setup), the build output is displayed correctly:
buildOutput

**Comment by ole-ve** [comment]
> Tested on TS2, creation of the gradle project works as expected. However, I noticed that no build output is available in case of a failed build.

Thanks for mentioning this issue. Unfortunately, I currently don't have time to work on this issue. If someone else want to work on the issue, this is what needs to be changed:

The source for the error is the way the build log is filtered for relevant lines in [JenkinsBuildLogParseUtils](https://github.com/ls1intum/Artemis/blob/develop/src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsBuildLogParseUtils.java)::parseBuildLogsFromJenkinsLogs. The implementation works fine for Maven build log, because every line starts with `[<timestamp>]`. Unfortunately, this does not apply to Gradle build log.

Example for Maven log:
`[2021-05-10T15:19:49.741Z] [ERROR] BubbleSort.java:[15,9] not a statement`

Example for Gradle log:
`
/var/jenkins_home/workspace/GITLABJENKINSPROG3/GITLABJENKINSPROG3-BASE/assignment/src/de/tum/in/ase/Client.java:8: error: <identifier> expected builderror
`

Example for what should be filtered out:
`cp: cannot stat 'build/test-results/test/*.xml': No such file or directory`

File: src/main/resources/templates/jenkins/java/gradle/regularRuns/Jenkinsfile-staticCodeAnalysis [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+                steps {
+                    checkout([$class: 'GitSCM', branches: [[name: '**']], doGenerateSubmoduleConfigurations: false, extensions: [], submoduleCfg: [], userRemoteConfigs: [[credentialsId: '#gitCredentials', name: 'tests', url: '#testRepository']]])
+                    dir('#assignmentCheckoutPath') {
+                        checkout([$class: 'GitSCM', branches: [[name: '**']], doGenerateSubmoduleConfigurations: false, extensions: [], submoduleCfg: [], userRemoteConfigs: [[credentialsId: '#gitCredentials', name: 'assignment', url: '#assignmentRepository']]])
+                    }
+                }
+            }
+            stage('Build') {
+                steps {
+                    sh './gradlew clean test'
```
Comment (b-fein) [inline src/main/resources/templates/jenkins/java/gradle/regularRuns/Jenkinsfile-staticCodeAnalysis:24]:
```suggestion
                    timestamps {
                        sh './gradlew clean test'
                    }
```

File: src/main/resources/templates/jenkins/java/gradle/sequentialRuns/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                steps {
+                    checkout([$class: 'GitSCM', branches: [[name: '**']], doGenerateSubmoduleConfigurations: false, extensions: [], submoduleCfg: [], userRemoteConfigs: [[credentialsId: '#gitCredentials', name: 'tests', url: '#testRepository']]])
+                    dir('#assignmentCheckoutPath') {
+                        checkout([$class: 'GitSCM', branches: [[name: '**']], doGenerateSubmoduleConfigurations: false, extensions: [], submoduleCfg: [], userRemoteConfigs: [[credentialsId: '#gitCredentials', name: 'assignment', url: '#assignmentRepository']]])
+                    }
+                }
+            }
+            stage('Build') {
+                steps {
+                    sh './gradlew clean structuralTest behaviorTest'
```
Comment (b-fein) [inline src/main/resources/templates/jenkins/java/gradle/sequentialRuns/Jenkinsfile:24]:
```suggestion
                    timestamps {
                        sh './gradlew clean structuralTest behaviorTest'
                    }
```

**Review by b-fein — COMMENTED** [review]
> The implementation works fine for Maven build log, because every line starts with [<timestamp>]. Unfortunately, this does not apply to Gradle build log.

This can be fixed as per the inline comments. The mechanism is identical to the one for the [Maven exercises](https://github.com/ls1intum/Artemis/blob/6ec8c5c6f257c7ef7229e379650148f0522d9792/src/main/resources/templates/jenkins/java/regularRuns/Jenkinsfile#L24).
The errors probably still don’t show up as annotations in the code editor, but maybe that is enough to show them in the log below the code editor. (I have not yet tested it.)

File: src/main/resources/templates/jenkins/java/gradle/regularRuns/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                steps {
+                    checkout([$class: 'GitSCM', branches: [[name: '**']], doGenerateSubmoduleConfigurations: false, extensions: [], submoduleCfg: [], userRemoteConfigs: [[credentialsId: '#gitCredentials', name: 'tests', url: '#testRepository']]])
+                    dir('#assignmentCheckoutPath') {
+                        checkout([$class: 'GitSCM', branches: [[name: '**']], doGenerateSubmoduleConfigurations: false, extensions: [], submoduleCfg: [], userRemoteConfigs: [[credentialsId: '#gitCredentials', name: 'assignment', url: '#assignmentRepository']]])
+                    }
+                }
+            }
+            stage('Build') {
+                steps {
+                    sh './gradlew clean test'
```
Comment (b-fein) [inline src/main/resources/templates/jenkins/java/gradle/regularRuns/Jenkinsfile:24]:
```suggestion
                    timestamps {
                        sh './gradlew clean test'
                    }
```

**Comment by ole-ve** [comment]
> > The implementation works fine for Maven build log, because every line starts with [<timestamp>]. Unfortunately, this does not apply to Gradle build log.
> 
> This can be fixed as per the inline comments. The mechanism is identical to the one for the [Maven exercises](https://github.com/ls1intum/Artemis/blob/6ec8c5c6f257c7ef7229e379650148f0522d9792/src/main/resources/templates/jenkins/java/regularRuns/Jenkinsfile#L24). The errors maybe still don’t show up as annotations in the code editor, but maybe that is enough to show them in the log below the code editor. (I have not yet tested it.)
> 
> Edit: I remembered that I even reviewed and tested that the annotations work in the Bamboo setup for Gradle exercises. The screenshots confirm that, so they should work here as well as the error message after the timestamp is the same as in Bamboo.

Thanks for the solution, it works as expected:

Bildschirmfoto 2022-04-15 um 13 42 01

**Review by dfuchss — APPROVED** [review]
Changes since last review look good! Thanks for your work!

**Pull request merged**